---
title: "ARVIN : Identifying Risk Noncoding Variants Using Disease-relevant Gene Regulatory Networks"
author: "Long Gao, Yasin Uzun, Kai Tan"
date: "October 31, 2017"
output: 
  pdf_document:
    toc: true
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1 Introduction

Identifying causalnoncoding variants remains a daunting task. Because noncoding variants exert their effectsin the context of a gene regulatory network (GRN), we hypothesize that explicit use of disease-relevant GRNcan significantly improve the inference accuracy of noncoding risk variants. We describe Annotation of Regulatory Variants using Integrated Networks (ARVIN), a general computational framework for predictingcausal noncoding variants. For each disease, ARVIN first constructs a GRN using multi-dimensional omics data oncell/tissue-type relevant to the disease. ARVIN then uses a set of novel regulatory network-basedfeatures, combined with sequence-based features to make predictions.This user guide contains the information necessary to run ARVIN.

## 2 Network construction
### 2.1 Enhancer prediction
ARVIN uses enhancer regions for mapping SNPs to enhancers. The data file must be in the following tab separated format:  
#chromosome	enhancer_center	enhancer_probability  
chr6	138231000	0.86  
chr1	160807000	0.79  
chr1	160808600	0.88  
You can use CSI-ANN software for predicting the enhancers. For this purpose, you can download CSI-ANN from our lab web page http://tanlab4generegulation.org/CSIANNWebpage.html . CSI-ANN uses histone modification data as input to predict enhancer regions in genome-wide. You can use the output of CSI-ANN (which is as shown above) as input for ARVIN.
### 2.2 Enhancer-promoter interaction prediction
ARVIN uses enhancer-promoter interaction for mapping SNPS to genes via enhancers. The enhancer-promoter interaction data must be in tab separated format as follows:  
#Chr	Start	End	Target	Score  
chr9	22124001	22126001	ENST00000452276	0.93  
chr2	242792001	242794001	ENST00000485966	0.792  
You can use IM-PET software for predicting enhancer-promoter interactions. For this purpose, you can download IM-PET from http://tanlab4generegulation.org/IM-PET.html . IM-PET uses enhancer predictions (computed using CSI-ANN) and gene expression data to compute enhancer-promoter interactions. You can use the output of IM-PET (which is as shown above) as input for ARVIN.

### 2.3 Obtain gene-gene interaction network
### 2.3 Network scoring

## 3 Prepare features for risk variants prediction
### 3.1 Network-based features
#### 3.1.1 Betweenness centrality
#### 3.1.2 Closeness centrality
#### 3.1.3 Pagerank centrality
#### 3.1.4 Weighted degree
#### 3.1.5 Module score
### 3.2 GWAVA features
ARVIN uses sequence features for the input SNPs generated by GWAVA. GWAVA is an open-source software developed by Sanger Institute. You can either upload the SNPs to GWAVA web page and get the output or download the source and run locally.
For running GWAVA online navigate to  https://www.sanger.ac.uk/sanger/StatGen_Gwava, upload the list of input SNPs and get the features in csv format, which will be input for ARVIN.
If you prefer to run it locally, you need to dowload the source code from ftp://ftp.sanger.ac.uk/pub/resources/software/gwava/v1.0/src/  and annotation data from ftp://ftp.sanger.ac.uk/pub/resources/software/gwava/v1.0/source_data/ . Then you can run it local by running gwava_annotate.py and generate the features, which will be input for ARVIN.
### 3.3 FunSeq features
ARVIN also uses sequence features generated by FunSeq. FunSeq can also be run online or binaries can be downloaded to run locally. 
For running FunSeq online, navigate to http://funseq.gersteinlab.org/analysis and upload the list of SNPs that you want to analyze. In the web page, it is noted that the input SNPs can be uploaded in bed format, SNP coordinates followed by reference and alternate alleles; but we discovered that it fails to process bed input. In order to have it run, you the first two seperators need to be two spaces and last two separators need to be tabs, as follows:  
chr16··4526757··4526758	G	A  
chr14··52733136··52733137	C	A  
where each dot (·) represents a space. 
Then, FunSeq will generate the features by selecting “bed” as the output format., which will be used as input by ARVIN.

If you prefer to run FunSeq locally, you can download FunSeq binaries from http://funseq.gersteinlab.org/static/funseq-0.1.tar.gz and extract it into your local. You will also need to download FunSeq annotation data from http://funseq.gersteinlab.org/static/data/data.tar.gz , extract it into directory that you saved the binaries. Then you can run FunSeq binary file by setting the output format to bed.  

## 4 Build a classifier for prioritizing risk varints
### 4.1 Train a random forest classifier
### 4.2 Predict causal disease variants
